language/vm/src/views.rs

```diff
@@ -531,63 +532,120 @@ impl<'a, T: ModuleAccess> SignatureTokenView<'a, T> {
+                //     - `all` means part of the type can be either `resource` or `unrestricted`.
+                //     - Therefore it is also impossible to determine the kind of the type as a
+                //       whole, and thus `all`.
+                //   - If none of the type actuals has kind `all`, then the struct is a resource if
+                //     and only if one of the type actuuals has kind `resource`.
+                let kind = kinds
+                    .iter()
+                    .fold(Kind::Unrestricted, |acc_kind, next_kind| {
+                        match (acc_kind, next_kind) {
+                            (Kind::All, _) | (_, Kind::All) => Kind::All,
+                            (Kind::Resource, _) | (_, Kind::Resource) => Kind::Resource,
+                            (Kind::Unrestricted, Kind::Unrestricted) => Kind::Unrestricted,
+                        }
+                    });
+
+                Ok(kind)
             }
-            SignatureToken::Reference(_)
-            | SignatureToken::MutableReference(_)
-            | SignatureToken::Bool
-            | SignatureToken::U64
-            | SignatureToken::String
-            | SignatureToken::ByteArray
-            | SignatureToken::Address => Kind::Unrestricted,
-            // TODO: To get the kind of a type parameter we need to look at the struct/function
-            // that contains it. Change the API or remodel accesses/views with a tiered system.
-            SignatureToken::TypeParameter(_) => panic!("cannot tell if a type parameter is a resource or not (feature not yet implemented)"),
         }
     }
 
-    pub fn contains_nominal_resource(&self) -> bool {
+    // TODO: refactor views so that we get the type formals from self.
+    pub fn contains_nominal_resource(&self, type_formals: &[Kind]) -> bool {
         match self.token {
-            // TODO: Type actuals are ignored, fix it (generics).
             SignatureToken::Struct(sh_idx, type_arguments) => {
                 StructHandleView::new(self.module, self.module.struct_handle_at(*sh_idx))
                     .is_nominal_resource()
-                    || type_arguments
-                        .iter()
-                        .any(|token| Self::new(self.module, token).contains_nominal_resource())
+                    || type_arguments.iter().any(|token| {
+                        Self::new(self.module, token).contains_nominal_resource(type_formals)
+                    })
             }
             SignatureToken::Reference(_)
             | SignatureToken::MutableReference(_)
             | SignatureToken::Bool
             | SignatureToken::U64
             | SignatureToken::String
             | SignatureToken::ByteArray
-            | SignatureToken::Address
-            | SignatureToken::TypeParameter(_) => false,
+            | SignatureToken::Address => false,
+
+            SignatureToken::TypeParameter(idx) => match type_formals[*idx as usize] {
```

Comment:
The other weird thing is that the parameter is called type_formals but an entry in it is actually a kind.  Perhaps kind_actuals might be a better name?

Reply:
Had a few discussions with @sblackshear on this. I can definitely see your confusion: the formals contain kind constraints and they are stored as an array of kinds in the file format. 

However I'd argue that the "actual" & "formal" are in terms of the "value" you feed in. For the following function call
```
foo(x: u64, y: bool) { ... }
// "x: u64, y: bool" is what we call "formals" right now
// it is stored as a `Vec<SignatureToken>` right now
// my understanding is that "formals" is an abbreviation of "value_formals" here

foo(123, true);
// "123, true" are what we call actuals. they are values
```
Similarly for a generic function
```
foo<T>(x: T) { ... }

foo<u64>(42);
```
The actuals are types, and hence it is more consistent to call the corresponding formals "type_formals" rather than "kind_formals". This becomes less obvious if you look at its representation in the file format: the named type variables are replaced with indices so you don't realize they still exist conceptually.